FreeRTOS/Demo/ARC_NSIM_QEMU_GCC_MW/arc/arc_exception.c

```diff
@@ -0,0 +1,260 @@
+
+	int_cause_reg = arc_aux_read(AUX_IRQ_CAUSE);
+	arc_kflag(1);
+}
+
+__attribute__ ((aligned(1024), section(".vector")))
+EXC_ENTRY_T exc_entry_table[NUM_EXC_ALL] = {
+	[0] = _arc_reset,
+	[1 ... NUM_EXC_CPU - 1] = exc_entry_cpu,
+	[NUM_EXC_CPU ... NUM_EXC_ALL - 1] = exc_entry_int
+};
+/**
+ * @var exc_int_handler_table
+ * @brief the cpu exception and interrupt exception handler table
+ * called in exc_entry_default and exc_entry_int
+ */
+EXC_HANDLER_T exc_int_handler_table[NUM_EXC_ALL] = {
+	[0 ... NUM_EXC_CPU - 1] = exc_handler_default,
+	[NUM_EXC_CPU ... NUM_EXC_ALL - 1] = int_handler_default
+};
+
+typedef struct aux_irq_ctrl_field {
+	/* note: little endian */
+	uint32_t save_nr_gpr_pairs : 5;         /** Indicates number of general-purpose register pairs saved, from 0 to 8/16 */
+	uint32_t res : 4;                       /** Reserved */
+	uint32_t save_blink : 1;                /** Indicates whether to save and restore BLINK */
+	uint32_t save_lp_regs : 1;              /** Indicates whether to save and restore loop registers (LP_COUNT, LP_START, LP_END) */
+	uint32_t save_u_to_u : 1;               /** Indicates if user context is saved to user stack */
+	uint32_t res2 : 1;                      /** Reserved */
+	uint32_t save_idx_regs : 1;             /** Indicates whether to save and restore code-density registers (EI_BASE, JLI_BASE, LDI_BASE) */
+	uint32_t res3 : 18;                     /** Reserved */
+} aux_irq_ctrl_field_t;
+
+typedef union {
+	aux_irq_ctrl_field_t bits;
+	uint32_t value;
+} aux_irq_ctrl_t;
+
+/**
+ * @ingroup ARC_HAL_EXCEPTION_CPU ARC_HAL_EXCEPTION_INTERRUPT
+ * @brief  initialize the exception and interrupt handling
+ */
+void exc_int_init(void)
+{
+	uint32_t i;
+	uint32_t status;
+	aux_irq_ctrl_t ictrl;
+
+	ictrl.value = 0;
+
+	ictrl.bits.save_nr_gpr_pairs = 6;       /* r0 to r11 (r12 saved manually) */
+
+	ictrl.bits.save_blink = 1;
+	ictrl.bits.save_lp_regs = 1;            /* LP_COUNT, LP_START, LP_END */
+	ictrl.bits.save_u_to_u = 0;             /* user ctxt saved on kernel stack */
+
+	status = arc_lock_save();
+	for (i = NUM_EXC_CPU; i < NUM_EXC_ALL; i++) {
+		/* interrupt level triggered, disabled, priority is the lowest */
+		arc_aux_write(AUX_IRQ_SELECT, i);
+		arc_aux_write(AUX_IRQ_ENABLE, 0);
+		arc_aux_write(AUX_IRQ_TRIGGER, 0);
+		arc_aux_write(AUX_IRQ_PRIORITY, INT_PRI_MAX - INT_PRI_MIN);
+	}
+	arc_aux_write(AUX_IRQ_CTRL, ictrl.value);
+
+	arc_unlock_restore(status);
+
+	/** ipm should be set after cpu unlock restore to avoid reset of the status32 value */
+	arc_int_ipm_set((INT_PRI_MAX - INT_PRI_MIN));
+}
+
+/**
+ * @ingroup ARC_HAL_EXCEPTION_CPU
+ * @brief  install an exception handler
+ * @param[in] excno	exception number
+ * @param[in] handler the handler of exception to install
+ */
+int32_t exc_handler_install(const uint32_t excno, EXC_HANDLER_T handler)
+{
+	if (excno < NUM_EXC_ALL && handler != NULL) {
+		exc_int_handler_table[excno] = handler;
+		return 0;
+	}
+
+	return -1;
+}
```

Comment:
Not so good demo as it will only work in simulation. On real HW (with caches) it will require D cache flush & I cache invalidate. 

Do we use it outside of this demo? Let's just put handlers in IVT statically if it's possible here.